File: Leetcode/linkedlist/24_SwapNodesinPairs/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head==null || head.next==null) return head;
        ListNode res = head.next, next = head.next.next;
        res.next = head;
        head.next = swapPairs(next);
        return res;
    }
}

File: Leetcode/linkedlist/24_SwapNodesinPairs/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(head==nullptr||head->next==nullptr) return head;
        ListNode *res = head->next, *nxt = head->next->next;
        res->next = head;
        head->next = swapPairs(nxt);
        return res;
    }
};
